File: src/main/java/no/stunor/origo/batch/model/Organisation.java
package no.stunor.origo.batch.model;

import com.google.cloud.Timestamp;
import com.google.cloud.firestore.annotation.DocumentId;
import com.google.cloud.spring.data.firestore.Document;

import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;

@Getter
@Setter
@Builder
@NoArgsConstructor
@AllArgsConstructor
@Document(collectionName = "organisations")
public class Organisation {
    @DocumentId
    private String id;
    private String organisationId;
    private String eventorId;
    private String name;
    private String contactPerson;
    private String email;
    private String type;
    private String regionId;
    private String country;
    private Timestamp lastUpdated;

    @Override
    public boolean equals(Object o) {
        if (o instanceof Organisation) {
            Organisation org = (Organisation) o;
            return this.eventorId.equals(org.eventorId) && this.organisationId.equals(org.organisationId);
        }
        return false;
    }
}

File: src/main/kotlin/no/stunor/origo/batch/model/Organisation.kt
package no.stunor.origo.batch.model

import jakarta.persistence.*
import java.sql.Timestamp
import java.time.Instant
import java.util.UUID

@Entity
data class Organisation (
    @Id @GeneratedValue(strategy = GenerationType.AUTO)
    var id: UUID? = null,
    var eventorId: String = "",
    var eventorRef: String = "",
    var name: String = "",
    @Enumerated(EnumType.STRING) var type: OrganisationType = OrganisationType.Club,
    var country: String = "",
    var email: String? = null,
    var eventorApiKey: String? = null,
    var regionId: UUID? = null,
    var contactPerson: String? = null,
    var lastUpdated: Timestamp = Timestamp.from(Instant.now())
) {
    override fun equals(other: Any?): Boolean {
        if (other is Organisation) {
            return this.eventorId == other.eventorId && (this.eventorRef == other.eventorId)
        }
        return false
    }

    fun isUpdatedAfter(other: Organisation): Boolean {
        return this.lastUpdated.after(other.lastUpdated)
    }
    override fun hashCode(): Int {
        var result = eventorId.hashCode()
        result = 31 * result + eventorRef.hashCode()
        result = 31 * result + name.hashCode()
        result = 31 * result + type.hashCode()
        result = 31 * result + country.hashCode()
        result = 31 * result + (email?.hashCode() ?: 0)
        result = 31 * result + (eventorApiKey?.hashCode() ?: 0)
        result = 31 * result + (regionId?.hashCode() ?: 0)
        result = 31 * result + (contactPerson?.hashCode() ?: 0)
        return result
    }
}

enum class OrganisationType {
    Club, Region, Federation, IOF
}
